src/Corax/Indexing/IndexWriter.FieldBuffers.cs

```diff
@@ -0,0 +1,138 @@
+            
+            if (_sortedTerms == null || _sortedTerms.Length < termsCount)
+            {
+                if (_sortedTerms != null)
+                {
+                    ArrayPool<TKey>.Shared.Return(_sortedTerms);
+                    ArrayPool<int>.Shared.Return(_termIndexes);
+                }
+
+                _sortedTerms = ArrayPool<TKey>.Shared.Rent(termsCount);
+                _termIndexes = ArrayPool<int>.Shared.Rent(termsCount);
+            }
+
+            int idx = 0;
+            if (typeof(TKey) == typeof(Slice))
+            {
+                foreach (var (k, v) in field.Textual)
+                {
+                    _sortedTerms[idx] = (TKey)(object)k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+            }
+            if (typeof(TKey) == typeof(long))
+            {
+                foreach (var (k, v) in field.Longs)
+                {
+                    _sortedTerms[idx] = (TKey)(object)k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+            }
+            if (typeof(TKey) == typeof(double))
+            {
+                foreach (var (k, v) in field.Doubles)
+                {
+                    _sortedTerms[idx] = (TKey)(object)k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+            }
+
+            terms = new Span<TKey>(_sortedTerms, 0, termsCount);
+            indexes = new Span<int>(_termIndexes, 0, termsCount);
+
+            if (typeof(TKey) == typeof(Slice))
+                (MemoryMarshal.Cast<TKey, Slice>(terms)).Sort(indexes, SliceComparer.Instance);
+            else if (typeof(TKey) == typeof(long))
+                (MemoryMarshal.Cast<TKey, long>(terms)).Sort(indexes);
+            else if (typeof(TKey) == typeof(double))
+                (MemoryMarshal.Cast<TKey, double>(terms)).Sort(indexes);
+            else
+                throw new InvalidDataException($"Type {typeof(TKey).FullName} is not supported");
+        }
+
+        public FieldBuffers(IndexWriter parent)
+        {
+            _parent = parent;
+            Keys = ArrayPool<TLookupKey>.Shared.Rent(BatchSize);
+            PageOffsets = ArrayPool<int>.Shared.Rent(BatchSize);
+            PostListIds = ArrayPool<long>.Shared.Rent(BatchSize);
+            _entriesOffsets = ArrayPool<int>.Shared.Rent(BatchSize);
+        }
+
+        public void Dispose()
+        {
+            if (PostListIds != null) ArrayPool<long>.Shared.Return(PostListIds);
+            if (PageOffsets != null) ArrayPool<int>.Shared.Return(PageOffsets);
+            if (_entriesOffsets != null) ArrayPool<int>.Shared.Return(_entriesOffsets);
+
+            if (_sortedTerms != null) ArrayPool<TKey>.Shared.Return(_sortedTerms);
+            if (_termIndexes != null) ArrayPool<int>.Shared.Return(_termIndexes);
+
```

Comment:
I dont really like those many `ArrayPool<T>` shared calls. We had been having to remove many of those because they act as a contention place over multiple unrelated codepaths and impact performance in pretty unreliable ways. If possible I would like to have `ContextBoundNativeList<T>` if possible instead.

Reply:
https://github.com/maciejaszyk/ravendb/tree/NumericalOptimizationsNative

I've created POC, I will let you know about the results.